tt_metal/impl/buffers/buffer.cpp

```diff
@@ -198,89 +227,158 @@ BufferPageMapping generate_buffer_page_mapping(const Buffer &buffer) {
     return buffer_page_mapping;
 }
 
-Buffer::Buffer(const Buffer &other) :
-    device_(other.device_),
-    size_(other.size_),
-    page_size_(other.page_size_),
-    buffer_type_(other.buffer_type_),
-    buffer_layout_(other.buffer_layout_),
-    shard_parameters_(other.shard_parameters_),
-    bottom_up_(other.bottom_up_),
-    buffer_page_mapping_(other.buffer_page_mapping_),
-    allocate_(other.allocate_) {
-    if (this->allocate_) {
-        this->allocate();
+void Buffer::allocate() {
+    {
+        std::unique_lock lock(allocation_mutex_);
+        TT_FATAL(allocation_status_.load(std::memory_order::relaxed) == AllocationStatus::NOT_ALLOCATED, "Can't allocate buffer after it was already allocated");
```

Comment:
there are two consecutive to the atomic variable. Something bad may happen in between.
But this code stinks in my opinion. I feel like  allocation_status_ should not be atomic because you need mutex anyway.

Reply:
I only see one query in the main thread here.

Reply:
@tt-asaigal The lock is crucial here, otherwise we would have a race condition. As @dmakoviichuk-tt pointed out, we're doing two consecutive calls to an atomic, and something can happen in between. Moreover, even if we make it a single call with `compare_exchange_strong` it would still be a race condition if someone calls `Buffer::address()` concurrently.  

In general, it's almost never correct to guard just CV with a mutex. See the first answer here: https://stackoverflow.com/questions/36126286/using-stdcondition-variable-with-atomicbool

@dmakoviichuk-tt I'm using an atomic here so we can avoid mutex locks in some places, for example in `Buffer::is_allocated()`. I think it might be used quite often, so having an atomic with all access being std::memory_order::relaxed likely saves some performance.